Sources/NativeBuilder/ContainerBuildSnapshotter/DirectoryDiffer.swift

```diff
@@ -0,0 +1,401 @@
+                    let newAttrs = try await inspector.inspect(url: newURL, options: options)
+                    if newAttrs.type == .characterDevice || newAttrs.type == .blockDevice {
+                        return nil
+                    }
+                    return nil
+                }
+                inFlight += 1
+                if inFlight >= concurrency {
+                    if let d = try await group.next(), let dd = d { changes.append(dd) }
+                    inFlight -= 1
+                }
+            }
+            while let d = try await group.next() {
+                if let dd = d { changes.append(dd) }
+            }
+        }
+
+        return dedupeStable(changes).sorted(by: diffPathLessThan)
+    }
+
+    private func buildFileMap(for directory: URL) async throws -> [BinaryPath: NormalizedFileAttributes] {
+        var map: [BinaryPath: NormalizedFileAttributes] = [:]
+
+        guard
+            let enumerator = FileManager.default.enumerator(
+                at: directory,
+                includingPropertiesForKeys: [
+                    URLResourceKey.isRegularFileKey,
+                    URLResourceKey.isDirectoryKey,
+                    URLResourceKey.isSymbolicLinkKey,
+                    URLResourceKey.fileSizeKey,
+                    URLResourceKey.contentModificationDateKey,
+                ],
+                options: []
+            )
+        else {
+            throw DifferError.cannotEnumerateDirectory(directory)
+        }
+
+        while let fileURL = enumerator.nextObject() as? URL {
+            let relativePath = self.relativePath(of: fileURL, base: directory)
+            var attrs = try await inspector.inspect(url: fileURL, options: options)
+            // Preserve canonical relative path in attributes for potential downstream use
+            attrs.path = relativePath
+            map[relativePath] = attrs
+        }
+
+        return map
+    }
+
+    // MARK: - Helpers
+
+    private func relativePath(of url: URL, base: URL) -> BinaryPath {
+        // Standardize both URLs to handle macOS /private prefix
+        let standardURL = url.standardizedFileURL
+        let standardBase = base.standardizedFileURL
+
+        // Get raw bytes from standardized filesystem representations
+        let standardURLPath = BinaryPath(url: standardURL)
+        let standardBasePath = BinaryPath(url: standardBase)
+
+        // Try to compute relative path using standardized paths
+        if let relative = standardURLPath.relativePath(from: standardBasePath) {
+            return relative
+        }
+
+        // Fallback: try with original paths
+        let urlPath = BinaryPath(url: url)
+        let basePath = BinaryPath(url: base)
+
+        if let relative = urlPath.relativePath(from: basePath) {
+            return relative
+        }
+
+        // Last resort - shouldn't happen in normal usage
+        return urlPath.lastPathComponent
```

Comment:
When would these happen?

Reply:
This is due to swift URL's non-deterministic behavior

- The first call to relativePath computes relative path based on standardizedFileURL 
 - The second call to relativePath computes relative path based on direct URL (without calling standardizedFileURL)
- If both of these do not yield a relative path, then this acts as a safe fallback. This can happen if `standardizedFIleURL` does not work as expected (which we've run into before)

Worst case, it leads to a key being counted as a .modified or .added for a particular path.